### PR TITLE
iOS: add IOS delegate, rect, userAgent and evalJavascript

### PR DESCRIPTION
**Pull request by pedia**
add UIWebViewDelegate for UIWebView
add rect/hidden argument for the first launch

### CODE REVIEW

File: lib/flutter_webview_plugin.dart [hunk 1]
```diff
@@ -1,33 +1,41 @@
...
-
-  _init() {
+  FlutterWebviewPlugin() : 
+    _channel = const MethodChannel(_kChannel),
+    _event = const EventChannel(_kEvent) {
     _channel.setMethodCallHandler(_handleMessages);
   }
 
   Future<Null> _handleMessages(MethodCall call) async {
+    print("_handleMessages $call");
```
Comment (lejard-h) [inline lib/flutter_webview_plugin.dart]: remove print

File: lib/flutter_webview_plugin.dart [hunk 2]
```diff
@@ -1,33 +1,41 @@
...
-  FlutterWebviewPlugin._() {
-    _init();
-  }
-
-  factory FlutterWebviewPlugin() => _instance ??= new FlutterWebviewPlugin._();
-
-  _init() {
+  FlutterWebviewPlugin() : 
+    _channel = const MethodChannel(_kChannel),
+    _event = const EventChannel(_kEvent) {
```
Comment (lejard-h) [inline lib/flutter_webview_plugin.dart]: why a new channel and not use the existing MethodChannel ?

File: lib/flutter_webview_plugin.dart [hunk 3]
```diff
@@ -1,33 +1,41 @@
...
 /// Singleton Class that communicate with a fullscreen Webview Instance
 /// Have to be instanciate after `runApp` called.
 class FlutterWebviewPlugin {
-  final MethodChannel _channel = const MethodChannel(_kChannel);
+  final MethodChannel _channel;
+
+  final EventChannel _event;
+  Stream<String> _stateChanged;
+
+  Stream<String> get stateChanged {
```
Comment (lejard-h) [inline lib/flutter_webview_plugin.dart]: onUrlChanged Stream already exist for this

File: ios/Classes/FlutterWebviewPlugin.m [hunk 4]
```diff
@@ -32,27 +49,97 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
...
+    NSURLRequest *request = [NSURLRequest requestWithURL:[NSURL URLWithString:url]];
+    [self.webview loadRequest:request];
+}
+
+- (void)initWebView:(FlutterMethodCall*)call {
+    // NSNumber *withJavascript = call.arguments[@"withJavascript"];
     NSNumber *clearCache = call.arguments[@"clearCache"];
     NSNumber *clearCookies = call.arguments[@"clearCookies"];
-    NSNumber *fullScreen = call.arguments[@"fullScreen"];
+    NSNumber *hidden = call.arguments[@"hidden"];
```
Comment (lejard-h) [inline ios/Classes/FlutterWebviewPlugin.m]: could you document this ? removing `fullscreen` is a breaking change and I need to know if I can do the same thing for Android

File: ios/Classes/FlutterWebviewPlugin.m [hunk 5]
```diff
@@ -32,27 +49,97 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
...
+- (void)launch:(FlutterMethodCall*)call {
     NSString *url = call.arguments[@"url"];
-    NSNumber *withJavascript = call.arguments[@"withJavascript"];
+    
+    NSURLRequest *request = [NSURLRequest requestWithURL:[NSURL URLWithString:url]];
+    [self.webview loadRequest:request];
+}
+
+- (void)initWebView:(FlutterMethodCall*)call {
+    // NSNumber *withJavascript = call.arguments[@"withJavascript"];
```
Comment (lejard-h) [inline ios/Classes/FlutterWebviewPlugin.m]: I have no idea, but can we deactivate/activate javascript for iOS webview ?

File: lib/flutter_webview_plugin.dart (hunk 3 above)
Comment (lejard-h) [reply lib/flutter_webview_plugin.dart]: If there is a difference could you document it ?

File: ios/Classes/FlutterWebviewPlugin.m (hunk 5 above)
Comment (pedia) [reply ios/Classes/FlutterWebviewPlugin.m]: I don't think it's easy.

File: ios/Classes/FlutterWebviewPlugin.m (hunk 4 above)
Comment (pedia) [reply ios/Classes/FlutterWebviewPlugin.m]: OK. I will add some document.

fullScreen is default.

File: lib/flutter_webview_plugin.dart (hunk 3 above)
Comment (pedia) [reply lib/flutter_webview_plugin.dart]: It is difficult to implement onUrlchanged, I will try this.

Comment (lejard-h) [reply lib/flutter_webview_plugin.dart]: so what it the purpose of stateChanged ?

**Comment by lejard-h** [comment]
@toufikzitouni do you have time to review the iOS part ?

File: lib/flutter_webview_plugin.dart (hunk 2 above)
Comment (pedia) [reply lib/flutter_webview_plugin.dart]: I don't know how to send event via MethodChannel. 
https://github.com/flutter/plugins/blob/master/packages/battery/lib/battery.dart

File: lib/flutter_webview_plugin.dart (hunk 1 above)
Comment (pedia) [reply lib/flutter_webview_plugin.dart]: My bad.

File: lib/flutter_webview_plugin.dart [hunk 6]
```diff
@@ -61,22 +92,43 @@ class FlutterWebviewPlugin {
...
+  /// - [clearCookies] clear all cookies of the Webview
+  ///     iOS WebView: Not implemented yet
+  ///     iOS WKWebView: will implement later
+  ///     android: Implemented
+  /// - [hidden] not show
+  ///     iOS WebView: not shown(addSubView) in ViewController
+  ///     android: Implemented
+  ///   [fullScreen]: show in full screen mode, default true
+  ///     iOS WebView: without rect, show in full screen mode
+  ///     android: Not implemented yet
```
Comment (lejard-h) [inline lib/flutter_webview_plugin.dart]: fullscreen is implemented on Android, but hidden is not

File: lib/flutter_webview_plugin.dart [hunk 7]
```diff
@@ -61,22 +92,43 @@ class FlutterWebviewPlugin {
...
+  ///   [rect]: show in rect(not full screen)
+  ///     iOS WebView: worked
+  ///     android: Not implemented yet
+  ///   [enableAppScheme]: false will enable all schemes, true only for
```
Comment (lejard-h) [inline lib/flutter_webview_plugin.dart]: iOS implemented,
Android not implemented

File: lib/flutter_webview_plugin.dart [hunk 8]
```diff
@@ -90,6 +142,10 @@ class FlutterWebviewPlugin {
     await _channel.invokeMethod('launch', args);
   }
 
+  Future<String> evalJavascript(String code) {
```
Comment (lejard-h) [inline lib/flutter_webview_plugin.dart]: IOS implemented
Android not implemented

File: lib/flutter_webview_plugin.dart [hunk 9]
```diff
@@ -28,14 +60,13 @@ class FlutterWebviewPlugin {
   final StreamController<String> _onUrlChanged =
       new StreamController.broadcast();
 
-  FlutterWebviewPlugin() : 
-    _channel = const MethodChannel(_kChannel),
-    _event = const EventChannel(_kEvent) {
+  FlutterWebViewPlugin()
+      : _channel = const MethodChannel(_kChannel),
```
Comment (lejard-h) [inline lib/flutter_webview_plugin.dart]: can't we use on `MethodChannel` instead of 2 channels ?

File: lib/flutter_webview_plugin.dart (hunk 3 above)
Comment (lejard-h) [reply lib/flutter_webview_plugin.dart]: can you add comment ?
iOS implemented, 
Android not implemented

Comment (pedia) [reply lib/flutter_webview_plugin.dart]: OK

File: lib/flutter_webview_plugin.dart (hunk 6 above)
Comment (pedia) [reply lib/flutter_webview_plugin.dart]: OK

File: lib/flutter_webview_plugin.dart (hunk 7 above)
Comment (pedia) [reply lib/flutter_webview_plugin.dart]: OK

File: lib/flutter_webview_plugin.dart (hunk 8 above)
Comment (pedia) [reply lib/flutter_webview_plugin.dart]: OK

**Comment by pedia** [comment]
@lejard-h 
I remove ios event channel, add rect, userAgent ... for android

**Comment by pedia** [comment]
@lejard-h 

Maybe I should close this pull request, and request a new one.

**Pull request closed without merging**